_posts/2025-05-09-opentelemetry-zero-code-instrumentation.md

```diff
@@ -0,0 +1,537 @@
+categories:
+- functions
+- otel
+- traces
+dark_background: true
+image: ""
+hide_header_image: true
+---
+
+Zero-code instrumentation for OpenFaaS function with OpenTelemetry
+
+Why would you instrument OpenFaaS functions?
+
+By collecting traces from functions, especially in more complex applications where functions are chained together or where they are calling a lot of external services traces can help you:
+
+- Identify performance bottlenecks
+- Detect errors and perform root cause analysis.
+- Understand system behaviour.
+
+Traces can be a great addition to the wide [range of Prometheus metrics](https://docs.openfaas.com/architecture/metrics/) OpenFaaS already makes available for monitoring functions.
+
+![](/images/2025-05-opentelemetry/function-otel-collection-diagram.png)
+
+In this tutorial we are going to:
+
+- Deploy [Grafana Tempo](https://grafana.com/oss/tempo/) as the backend for storing traces.
+- Run and configure [Grafana Alloy](https://grafana.com/oss/alloy-opentelemetry-collector/) as a [Collector](https://opentelemetry.io/docs/collector/) for traces.
+- Capture telemetry from functions using zero-code instrumentation.
+
+
+## Prepare the cluster for telemetry collection
+
+Before instrumenting functions we need to prepare our cluster for collecting and storing telemetry data.
+
+OpenTelemetry works by collecting telemetry data (metrics, logs, traces) from applications using SDKs or agents. This data is sent to an OpenTelemetry Collector, which processes and exports it to a backend storage system ([Jaeger](https://www.jaegertracing.io/), [Zipkin](https://zipkin.io/), [Prometheus](https://prometheus.io/) or [vendor-specific](https://opentelemetry.io/ecosystem/vendors/) backend) for analysis and visualization.
+
+When using OpenTelemetry, the [Collector](https://opentelemetry.io/docs/collector/) is a vendor-agnostic proxy that can receive, process and filter telemetry data before sending it to one or more backends for storage. We are going to deploy [Grafana Alloy](https://grafana.com/docs/alloy/latest/) as a Collector. As the storage backend for traces we will be using [Grafana Tempo](https://grafana.com/oss/tempo/). We will not be collecting logs or metrics in this tutorial.
+
+### Deploy Grafana Tempo
+
+The tracing backend will receive data from the application using the OpenTelemetry Protocol (OTLP) and store it to be queried and retrieved for inspection and visualization. We will deploy Grafana Tempo to our cluster using the Helm chart.
+
+Tempo uses Object storage to store its data. We are going to use MinIO as the storage backend which is included in the Helm chart for demonstration purposes only. When using Tempo in production you should configure a different storage option. Tempo supports S3 compatible storage backends like [AWS S3](https://aws.amazon.com/pm/serv-s3/), [Azure Blob storage](https://azure.microsoft.com/en-us/products/storage/blobs), [Google Cloud Storage](https://cloud.google.com/storage), etc.
```

Comment:
*Object Storage* seems like it should be captialised as a proper noun

Reply:
Make MinIO a link - put AGPL in brackets. Mention other solutions are available such as SeaweedFS (put its license in brackets)